Applications/Resnet/jni/main.cpp

```diff
@@ -65,23 +68,140 @@ static std::string withKey(const std::string &key,
+  LayerHandle a3 = create_conv("a3", 3, 1, 1, scoped_name("a2"));
+
+  /** skip path */
+  LayerHandle b1 = nullptr;
+  if (downsample) {
+    b1 = create_conv("b1", 1, 2, 0, input_name);
+  }
+
+  const std::string skip_name = b1 ? scoped_name("b1") : input_name;
+
+  LayerHandle c1 = createLayer(
+    "Addition",
+    {with_name("c1"), withKey("input_layers", {scoped_name("a3"), skip_name})});
+
+  LayerHandle c2 = create_batch_relu(""); /// use block_name itself.
+
+  if (downsample) {
+    return {a1, a2, a3, b1, c1, c2};
+  } else {
+    return {a1, a2, a3, c1, c2};
+  }
 }
 
-/**s
+/**
  * @brief Create resnet 18
  *
- * @return ModelHandle to create model
+ * @return vector of layers that contain full graph of resnet18
  */
-ModelHandle createResnet18() {
+std::vector<LayerHandle> createResnet18Graph() {
+  using ml::train::createLayer;
+
   std::vector<LayerHandle> layers;
-  return nullptr;
+
+  layers.push_back(
+    createLayer("conv2d", {
+                            withKey("name", "conv0"),
+                            withKey("kernel_size", {3, 3}),
+                            withKey("stride", {1, 1}),
+                            withKey("padding", {1, 1}),
+                            withKey("bias_initializer", "zeros"),
+                            withKey("weight_initializer", "xavier_uniform"),
+                          }));
+
+  layers.push_back(
+    createLayer("batch_normalization", {withKey("name", "first_bn_relu"),
+                                        withKey("activation", "relu")}));
+
+  std::vector<std::vector<LayerHandle>> blocks;
+
+  blocks.push_back(resnetBlock("conv1_0", "first_bn_relu", 64, 3, false));
+  blocks.push_back(resnetBlock("conv1_1", "conv1_0", 64, 3, false));
+  blocks.push_back(resnetBlock("conv2_0", "conv1_1", 128, 3, true));
+  blocks.push_back(resnetBlock("conv2_1", "conv2_0", 128, 3, false));
+  blocks.push_back(resnetBlock("conv3_0", "conv2_1", 256, 3, true));
+  blocks.push_back(resnetBlock("conv3_1", "conv3_0", 256, 3, false));
+  blocks.push_back(resnetBlock("conv4_0", "conv3_1", 256, 3, true));
+  blocks.push_back(resnetBlock("conv4_1", "conv4_0", 256, 3, false));
+
+  for (auto &block : blocks) {
+    layers.insert(layers.end(), block.begin(), block.end());
+  }
+
+  layers.push_back(createLayer("pooling2d", {withKey("name", "last_p1"),
+                                             withKey("pooling", "average"),
+                                             withKey("pool_size", {4, 4})}));
+
+  layers.push_back(createLayer("flatten", {withKey("name", "last_f1")}));
+  layers.push_back(
+    createLayer("fully_connected",
+                {withKey("unit", 100), withKey("activation", "softmax")}));
+
+  return layers;
+}
+
+ModelHandle createResnet18() {
+  ModelHandle model =
+    ml::train::createModel(ml::train::ModelType::NEURAL_NET,
+                           {withKey("loss", "cross"),
+                            withKey("batch_size", 128), withKey("epochs", 60)});
```

Comment:
You might want to reduce training times by a smaller number of epochs (possibly use pre-trained resnet18 for initialization) for a later PR.

Reply:
yes, I'll tune epochs when training